Use name for output variables instead of out in JIT

cpuhrsch opened the pull request:
This adds 88 matches.

cpuhrsch committed:
Use name for output variables instead of out in JIT

zdevito commented:
I am pretty sure that renaming line was necessary to make the `out` keyword-only argument work. It was added in this PR: https://github.com/pytorch/pytorch/pull/13093/files which adds support for that. It is possible it is no longer necessary if all the `_out` variants have renamed their output variable to `out`.

cpuhrsch commented:
@zdevito - ok, it looks like it's worth digging deeper into this.

cpuhrsch commented:
The generated code [changeset](https://gist.github.com/cpuhrsch/fe13cce8610b6e86b377e6ff05a18572) affects

```
        modified:   build/aten/src/ATen/Declarations.yaml
        modified:   torch/csrc/jit/generated/register_aten_ops_0.cpp
        modified:   torch/csrc/jit/generated/register_aten_ops_1.cpp
        modified:   torch/csrc/jit/generated/register_aten_ops_2.cpp
```

but doesn't impact any Python functions. Since the JIT doesn't care about the names of keyword arguments, it looks like these functions aren't exposed via Python anyway. All other functions for which it matters are thus presumably already using `out` as the output name as is required by the Python frontend.

cpuhrsch committed:
deal with add

cpuhrsch committed:
tmp

cpuhrsch committed:
[wip] also consider python bindings

cpuhrsch committed:
update yamls

cpuhrsch committed:
more yaml

cpuhrsch committed:
activate assert

cpuhrsch committed:
nn python bindings

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4525,119 +4589,136 @@
   python_module: nn
 
-- func: tanh_backward(Tensor grad_output, Tensor output, *, Tensor(a!) grad_input) -> Tensor(a!)
+- func: tanh_backward(Tensor grad_output, Tensor out, *, Tensor(a!) grad_input) -> Tensor(a!)
```
here too.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4525,119 +4589,136 @@
   python_module: nn
 
-- func: sigmoid_backward(Tensor grad_output, Tensor output, *, Tensor(a!) grad_input) -> Tensor(a!)
+- func: sigmoid_backward(Tensor grad_output, Tensor out, *, Tensor(a!) grad_input) -> Tensor(a!)
```
this looks weird -- output is an output in this case.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4525,119 +4589,136 @@
   python_module: nn
 
-- func: sigmoid_backward(Tensor grad_output, Tensor output) -> Tensor
+- func: sigmoid_backward(Tensor grad_output, Tensor out) -> Tensor
```
here too.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4525,119 +4589,136 @@
   python_module: nn
 
-- func: tanh_backward(Tensor grad_output, Tensor output) -> Tensor
+- func: tanh_backward(Tensor grad_output, Tensor out) -> Tensor
```
here too.

gchanan commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -449,7 +449,7 @@ def match_signature(decl, constructed_string, should_match_schema):
 def signature(decl, should_match_schema=True):
     def format_arg(arg):
-        name = arg['name'] if not arg.get('output') else 'out'
+        name = arg['name']  # if not arg.get('output') else 'out'
```
can you clean this up?

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -891,6 +891,14 @@ def get_py_formal_arg(arg):
+            assert output_args[0]['name'] == 'out', \
+                "For function named {}: Our Python frontend uses out as" \
+                " the keyword argument name for output arguments.".format(name)
+        # NB: For more than 1 output args the type name is a TensorList
```
is this note correct?  There are a bunch of cases with multiple output arguments that are Tensors AFAII.

cpuhrsch replied:
It's determined above. `typenames` is of the same length as `output_args` and `typename` is set to be `TensorList[len(typenames)]` if that's longer than 1.

cpuhrsch committed:
fix some arguments

cpuhrsch committed:
more python glue

cpuhrsch committed:
change some argument names

cpuhrsch commented:
[This](https://gist.github.com/cpuhrsch/2b7a2326070679b26a7e756088d7a1ba) is the codegen changeset.

gchanan commented on aten/src/ATen/native/README.md:
```diff
@@ -112,6 +112,12 @@ argument name you provide, and a rename of an argument name is considered a BC-b
 change (e.g., you will probably need to update `tools/autograd/derivatives.yaml` at
 least). For more details please see the section on `variants`.
 
+As a convention we use 'out' to indicate an output argument. This aligns with the
```
nit: this message is a little strange.  "This aligns with the Python bindings" is true, but the python bindings should be generated based on the schema here (and I think most would expect it is).

This also doesn't apply to operators with multiple output arguments.

"Check the generated code when making a change
to make sure you're not breaking the API when renaming an argument name of an
existing function."

this applies to any parameter you rename, not just output arguments.
